db/migrate/20170713173801_extract_to_proc_calls_in_templates.rb

```diff
@@ -0,0 +1,50 @@
+class ExtractToProcCallsInTemplates < ActiveRecord::Migration
+  def up
+    Template.descendants.each do |klass|
+      klass.unscoped.all.each do |template|
+        template.template = convert(template.template)
+        if template.template_changed?
+          # we need to skip validations so we use #update_attributes
+          template.update_attribute :template, template.template
+        end
+      end
+    end
+
+    Parameter.descendants.each do |klass|
+      klass.all.each do |parameter|
+        parameter.value = convert(parameter.value)
+        if parameter.value_changed?
+          # we need to skip validations so we use #update_attributes
+          parameter.update_attribute :value, parameter.value
+        end
+      end
+    end
+
+    LookupKey.descendants.each do |klass|
+      klass.all.each do |parameter|
+        parameter.default_value = convert(parameter.default_value.to_s)
+        if parameter.default_value_changed?
+          # we need to skip validations so we use #update_attributes
+          parameter.update_attribute :default_value, parameter.value
+        end
+      end
+    end
+
+    LookupValue.all.each do |parameter|
+      parameter.value = convert(parameter.value.to_s)
+      if parameter.value_changed?
+        # we need to skip validations so we use #update_attributes
+        parameter.update_attribute :value, parameter.value
+      end
+    end
+  end
+
+  def down
+    # audit can be used for reverting unwanted changes
+    say "skipping parameters macros migration"
+  end
+
+  def convert(content)
+    content.gsub(/<%(.*?)\(?&:([a-zA-Z0-9!?_]+)\)?(.*)%>/, '<%\1{ |item| item.\2 }\3%>')
```

Comment:
You can use lookahead syntax which is supported in Ruby 2.0+. We had an example in our old subnet parser which was recently replaced by new one: https://github.com/lzap/smart-proxy/blob/ce684575ba8777ac4edb68c6b0dfb40fd69f3e3d/modules/dhcp_isc/isc_file_parser.rb#L61-L73

This will work, but do we really need to change all templates? Since we have a locking mechanism now, we can simple deploy fixed ones and ask users to modify their clones in upgrade notes. At least we should do backups, this can completely ruin a production instance in case there is a bug. I feel like letting it to error out is much safer.